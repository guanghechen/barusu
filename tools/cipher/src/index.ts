export * from './util/buffer'
export * from './util/catalog'
export * from './util/cipher'
export * from './util/error'
export * from './util/event-bus'
export * from './util/io'
export * from './util/logger'
export * from './util/master'
export * from './util/secret'
