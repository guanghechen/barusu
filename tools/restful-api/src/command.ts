#! /usr/bin/env node
import { execCommand } from './index'


execCommand(process.argv)
