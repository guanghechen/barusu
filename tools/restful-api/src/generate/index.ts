import commander from 'commander'
import path from 'path'
import { SubCommand, SubCommandHook } from '../core/types/commander'
import {
  GenerateContextConfig,
  GlobalContextConfig,
} from '../core/types/context'
import {
  CmdOptions,
  GenerateCmdOptions,
  GlobalCmdOptions,
} from '../core/types/option'
import { logger } from '../core/util/logger'
import { RestfulApiToolGeneratorContext } from './context'
import { RestfulApiToolGenerator } from './generator'
export * from './context'
export * from './generator'


export type GenerateCommandAfterTriggeredHook = (
  projectDir: string,
  options: GenerateCmdOptions
) => Promise<void> | void

export type GenerateCommandBeforeStartHook = (
  generator: RestfulApiToolGenerator,
  context: RestfulApiToolGeneratorContext,
) => Promise<void> | void

export type GenerateCommandAfterCompletedHook = (
  generator: RestfulApiToolGenerator,
  context: RestfulApiToolGeneratorContext,
) => Promise<void> | void

export type GenerateCommandHook =
  | GenerateCommandAfterTriggeredHook
  | GenerateCommandBeforeStartHook
  | GenerateCommandAfterCompletedHook


/**
 * sub-command `generate`
 */
export class GenerateCommand implements SubCommand<'generate', GenerateCmdOptions> {
  public readonly name = 'generate'
  protected readonly afterTriggeredHooks: GenerateCommandAfterTriggeredHook[] = []
  protected readonly beforeStartHooks: GenerateCommandBeforeStartHook[] = []
  protected readonly afterCompletedHooks: GenerateCommandAfterCompletedHook[] = []

  public mount(
    program: commander.Command,
    globalCmdOptions: GlobalCmdOptions,
    calcGlobalContextConfig: (projectDir: string) => GlobalContextConfig,
    calcContextConfig: (
      projectDir: string,
      cmdOptions: CmdOptions,
      ignoredKeys?: (keyof CmdOptions)[],
    ) => GenerateContextConfig,
  ): void {
    const self = this
    const generateOptions = self.defaultOptions
    const coverGenerateOption = (key: keyof GenerateCmdOptions, value: any) => {
      if (value == null) return
      generateOptions[key] = { value, userSpecified: true } as any
    }

    // init sub-command generate
    program
      .command('generate <project-dir>')
      .alias('g')
      .option('-p, --tsconfig-path <tsconfigPath>', 'specify the location (absolute or relative to the projectDir) of typescript config file.')
      .option('-I, --ignore-missing-models', 'ignore missing model.')
      .option('--clean', 'clean schema folders before generate.')
      .action(async function (projectDir: string, options: GenerateCmdOptions) {
        logger.setName(self.name)

        // execute hook after action triggered
        for (const hook of self.afterTriggeredHooks) {
          await hook(projectDir, options)
        }

        coverGenerateOption('tsconfigPath', options.tsconfigPath)
        coverGenerateOption('ignoreMissingModels', options.ignoreMissingModels)
        coverGenerateOption('clean', options.clean)

        // debug logger
        logger.debug('projectDir:', projectDir)
        logger.debug('[generateOptions] tsconfigPath:', generateOptions.tsconfigPath)
        logger.debug('[generateOptions] ignoreMissingModels:', generateOptions.ignoreMissingModels)
        logger.debug('[generateOptions] clean:', generateOptions.clean)

        // eslint-disable-next-line no-param-reassign
        projectDir = path.resolve(globalCmdOptions.cwd.value, projectDir || '')
        const globalContextConfig = calcGlobalContextConfig(projectDir)
        const contextConfig: GenerateContextConfig = calcContextConfig(projectDir, {
          ...globalCmdOptions,
          ...generateOptions
        })

        // debug logger
        logger.debug('projectDir:', projectDir)
        logger.debug('[generateContextConfig] tsconfigPath:', contextConfig.tsconfigPath)
        logger.debug('[generateContextConfig] ignoreMissingModels:', contextConfig.ignoreMissingModels)
        logger.debug('[generateContextConfig] clean:', contextConfig.clean)
        logger.debug('[generateContextConfig] ignoredDataTypes:', contextConfig.ignoredDataTypes)
        logger.debug('[generateContextConfig] apiConfigPath:', contextConfig.apiConfigPath)
        logger.debug('[generateContextConfig] schemaRootPath:', contextConfig.schemaRootPath)
        logger.debug('[generateContextConfig] encoding:', contextConfig.encoding)
        logger.debug('[generateContextConfig] schemaArgs:', contextConfig.schemaArgs)
        logger.debug('[generateContextConfig] additionalCompilerOptions:', contextConfig.additionalCompilerOptions)

        const context = new RestfulApiToolGeneratorContext(projectDir, globalContextConfig, contextConfig)
        const generator = new RestfulApiToolGenerator(context)

        // execute hook before executing command
        for (const hook of self.beforeStartHooks) {
          await hook(generator, context)
        }

        await generator.generate()

        // execute hook after command completed
        for (const hook of self.afterCompletedHooks) {
          await hook(generator, context)
        }
      })
  }

  public onHook(type: SubCommandHook.AFTER_TRIGGERED, hook: GenerateCommandAfterTriggeredHook): this
  public onHook(type: SubCommandHook.BEFORE_START, hook: GenerateCommandBeforeStartHook): this
  public onHook(type: SubCommandHook.AFTER_COMPLETED, hook: GenerateCommandAfterCompletedHook): this
  public onHook(type: SubCommandHook, hook: GenerateCommandHook): this {
    switch (type) {
      case SubCommandHook.AFTER_TRIGGERED:
        this.afterTriggeredHooks.push(hook as GenerateCommandAfterTriggeredHook)
        break
      case SubCommandHook.BEFORE_START:
        this.beforeStartHooks.push(hook as GenerateCommandBeforeStartHook)
        break
      case SubCommandHook.AFTER_COMPLETED:
        this.afterCompletedHooks.push(hook as GenerateCommandAfterCompletedHook)
        break
    }
    return this
  }


  public get defaultOptions(): GenerateCmdOptions {
    const defaultOptions: GenerateCmdOptions = {
      tsconfigPath: {
        value: 'tsconfig.json',
        userSpecified: false,
      },
      ignoreMissingModels: {
        value: false,
        userSpecified: false,
      },
      clean: {
        value: false,
        userSpecified: false,
      },
    }
    return defaultOptions
  }
}
