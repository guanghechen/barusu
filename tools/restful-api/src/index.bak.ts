import program from 'commander'
import Router from 'koa-router'
import { optionMaster } from 'option-master'
import { version } from '@barusu/tool-restful-api/package.json'
import { Commander } from './_core/commander'
import { GenerateCommand } from './_generate'
import { InitCommand } from './_init'
import { ServeCommand } from './_serve'


// export
export * from './_core/types/api-item'
export * from './_core/types/api-item-group'
export * from './_core/types/api-config'
export * from './_core/types/commander'
export * from './_core/types/context'
export * from './_core/types/option'
export * from './_core/util/context-util'
export * from './_core/util/logger'
export * from './_core/util/type-util'
export * from './util/api-parser'
export * from './_core/commander'
export * from './_generate'
export * from './_serve'
export * from './_init'
export { Router }


/**
 * Create cli program.
 * If args is not null, use this args to execute the command
 * @param args
 */
export function execCommand(args?: string[], subCommands: {
  generate?: GenerateCommand,
  serve?: ServeCommand,
  init?: InitCommand,
} = {}): Commander {
  const commander = new Commander(
    version,
    program,
    optionMaster,
  )

  if (subCommands.generate == null) {
    // eslint-disable-next-line no-param-reassign
    subCommands.generate = new GenerateCommand()
  }
  if (subCommands.serve == null) {
    // eslint-disable-next-line no-param-reassign
    subCommands.serve = new ServeCommand()
  }
  if (subCommands.init == null) {
    // eslint-disable-next-line no-param-reassign
    subCommands.init = new InitCommand()
  }

  // register sub-command
  commander
    .registerCommand(subCommands.generate)
    .registerCommand(subCommands.serve)
    .registerCommand(subCommands.init)

  // run command
  if (args != null) {
    commander.run(args)
  }
  return commander
}
