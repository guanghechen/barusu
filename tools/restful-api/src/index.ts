import './env/constant'

export * from './command/generate'
export * from './command/init'
export * from './command/serve'
export * from './core/command'
export * from './core/generate/command'
export * from './core/generate/context'
export * from './core/generate/processor'
export * from './core/generate/util'
export * from './core/init/command'
export * from './core/init/context'
export * from './core/init/processor'
export * from './core/init/util'
export * from './core/option'
export * from './core/serve/command'
export * from './core/serve/context'
export * from './core/serve/middleware/access-log'
export * from './core/serve/middleware/serve-data-file'
export * from './core/serve/processor'
export * from './env/constant'
export * from './env/logger'
export * from './env/util'
export * from './types/api'
export * from './types/api-item'
export * from './types/api-item-group'
export * from './types/raw/api'
export * from './types/raw/api-item'
export * from './types/raw/api-item-group'
export * from './util/api-parser'
