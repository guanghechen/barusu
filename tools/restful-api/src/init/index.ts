import commander from 'commander'
import path from 'path'
import { SubCommand, SubCommandHook } from '../core/types/commander'
import { GlobalContextConfig, InitContextConfig } from '../core/types/context'
import {
  CmdOptions,
  GlobalCmdOptions,
  InitCmdOptions,
} from '../core/types/option'
import { logger, updateLogName } from '../core/util/logger'
import { RestfulApiToolInitiatorContext } from './context'
import { RestfulApiToolInitiator } from './initiator'
export * from './context'
export * from './initiator'


export type InitCommandAfterTriggeredHook = (
  projectDir: string,
  options: InitCmdOptions
) => Promise<void> | void

export type InitCommandBeforeStartHook = (
  initiator: RestfulApiToolInitiator,
  context: RestfulApiToolInitiatorContext,
) => Promise<void> | void

export type InitCommandAfterCompletedHook = (
  initiator: RestfulApiToolInitiator,
  context: RestfulApiToolInitiatorContext,
) => Promise<void> | void

export type InitCommandHook =
  | InitCommandAfterTriggeredHook
  | InitCommandBeforeStartHook
  | InitCommandAfterCompletedHook


/**
 * sub-command `init`
 */
export class InitCommand implements SubCommand<'init', InitCmdOptions> {
  public readonly name = 'init'
  protected readonly afterTriggeredHooks: InitCommandAfterTriggeredHook[] = []
  protected readonly beforeStartHooks: InitCommandBeforeStartHook[] = []
  protected readonly afterCompletedHooks: InitCommandAfterCompletedHook[] = []

  public mount(
    program: commander.Command,
    globalCmdOptions: GlobalCmdOptions,
    calcGlobalContextConfig: (projectDir: string) => GlobalContextConfig,
    calcContextConfig: (
      projectDir: string,
      cmdOptions: CmdOptions,
      ignoredKeys?: (keyof CmdOptions)[],
    ) => InitContextConfig,
  ): void {
    const self = this
    const initOptions = self.defaultOptions

    // init sub-command init
    program
      .command('init <project-dir>')
      .alias('i')
      .action(async function (projectDir: string, options: InitCmdOptions) {
        updateLogName(self.name)

        // execute hook after action triggered
        for (const hook of self.afterTriggeredHooks) {
          await hook(projectDir, options)
        }

        // debug logger
        logger.debug('projectDir:', projectDir)

        // eslint-disable-next-line no-param-reassign
        projectDir = path.resolve(globalCmdOptions.cwd.value, projectDir || '')
        const globalContextConfig = calcGlobalContextConfig(projectDir)
        const contextConfig: InitContextConfig = calcContextConfig(projectDir, {
          ...globalCmdOptions,
          ...initOptions
        }, [ 'apiConfigPath', 'schemaRootPath', 'encoding'])

        // debug logger
        logger.debug('projectDir:', projectDir)

        const context = new RestfulApiToolInitiatorContext(projectDir, globalContextConfig, contextConfig)
        const Initiator = new RestfulApiToolInitiator(context)

        // execute hook before executing command
        for (const hook of self.beforeStartHooks) {
          await hook(Initiator, context)
        }

        await Initiator.init()

        // execute hook after command completed
        for (const hook of self.afterCompletedHooks) {
          await hook(Initiator, context)
        }
      })
  }

  public onHook(type: SubCommandHook.AFTER_TRIGGERED, hook: InitCommandAfterTriggeredHook): this
  public onHook(type: SubCommandHook.BEFORE_START, hook: InitCommandBeforeStartHook): this
  public onHook(type: SubCommandHook.AFTER_COMPLETED, hook: InitCommandAfterCompletedHook): this
  public onHook(type: SubCommandHook, hook: InitCommandHook): this {
    switch (type) {
      case SubCommandHook.AFTER_TRIGGERED:
        this.afterTriggeredHooks.push(hook as InitCommandAfterTriggeredHook)
        break
      case SubCommandHook.BEFORE_START:
        this.beforeStartHooks.push(hook as InitCommandBeforeStartHook)
        break
      case SubCommandHook.AFTER_COMPLETED:
        this.afterCompletedHooks.push(hook as InitCommandAfterCompletedHook)
        break
    }
    return this
  }


  public get defaultOptions(): InitCmdOptions {
    const defaultOptions: InitCmdOptions = {
    }
    return defaultOptions
  }
}
