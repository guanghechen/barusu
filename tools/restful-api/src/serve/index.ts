import commander from 'commander'
import { CoverOperationFunc, coverInteger, coverString } from 'option-master'
import path from 'path'
import { SubCommand, SubCommandHook } from '../core/types/commander'
import { GlobalContextConfig, ServeContextConfig } from '../core/types/context'
import {
  CmdOptions,
  GlobalCmdOptions,
  ServeCmdOptions,
} from '../core/types/option'
import { logger } from '../core/util/logger'
import { RestfulApiToolServerContext } from './context'
import { RestfulApiToolServer } from './server'
export * from './context'
export * from './server'


export type ServeCommandAfterTriggeredHook = (
  projectDir: string, options: ServeCmdOptions
) => Promise<void> | void


export type ServeCommandBeforeStartHook = (
  server: RestfulApiToolServer,
  context: RestfulApiToolServerContext,
) => Promise<void> | void

export type ServeCommandAfterCompletedHook = (
  server: RestfulApiToolServer,
  context: RestfulApiToolServerContext,
) => Promise<void> | void

export type ServeCommandHook =
  | ServeCommandAfterTriggeredHook
  | ServeCommandBeforeStartHook
  | ServeCommandAfterCompletedHook


/**
 * sub-command `serve`
 */
export class ServeCommand implements SubCommand<'serve', ServeCmdOptions> {
  public readonly name = 'serve'
  protected readonly afterTriggeredHooks: ServeCommandAfterTriggeredHook[] = []
  protected readonly beforeStartHooks: ServeCommandBeforeStartHook[] = []
  protected readonly afterCompletedHooks: ServeCommandAfterCompletedHook[] = []

  public mount(
    program: commander.Command,
    globalCmdOptions: GlobalCmdOptions,
    calcGlobalContextConfig: (projectDir: string) => GlobalContextConfig,
    calcContextConfig: (
      projectDir: string,
      cmdOptions: CmdOptions,
      ignoredKeys?: (keyof CmdOptions)[],
    ) => ServeContextConfig,
  ): void {
    const self = this
    const serveOptions = self.defaultOptions
    const coverServeOption = (key: keyof ServeCmdOptions, value: any) => {
      if (value == null) return
      serveOptions[key] = { value, userSpecified: true } as any
    }

    // init sub-command serve
    program
      .command('serve <project-dir>')
      .alias('s')
      .option('-h, --host <host>', 'specify the ip/domain address to which the mock-server listens.')
      .option('-p, --port <port>', 'specify the port on which the mock-server listens.')
      .option('--prefix-url <prefixUrl>', 'specify the prefix url of routes.')
      .option('--mock-required-only', 'json-schema-faker\'s option `requiredOnly`')
      .option('--mock-optionals-always', 'json-schema-faker\'s option `alwaysFakeOptionals`')
      .option('--mock-optionals-probability <optionalsProbability>', 'json-schema-faker\'s option `optionalsProbability`')
      .option('--mock-use-data-file-first <mockDataFileRootPath>', 'specify the mock data file root path.')
      .option('--mock-data-file-first', 'preferred use data file as mock data source.')
      .action(async function (projectDir: string, options: ServeCmdOptions) {
        logger.setName(self.name)

        // execute hook after action triggered
        for (const hook of self.afterTriggeredHooks) {
          await hook(projectDir, options)
        }

        coverServeOption('host', options.host)
        coverServeOption('port', options.port)
        coverServeOption('mockDataFileRootPath', options.mockDataFileRootPath)
        coverServeOption('mockDataFileFirst', options.mockDataFileFirst)
        coverServeOption('mockRequiredOnly', options.mockRequiredOnly)
        coverServeOption('mockOptionalsAlways', options.mockOptionalsAlways)
        coverServeOption('mockOptionsProbability', options.mockOptionsProbability)

        // debug logger
        logger.debug('projectDir:', projectDir)
        logger.debug('[serveOptions] host:', serveOptions.host)
        logger.debug('[serveOptions] port:', serveOptions.port)
        logger.debug('[serveOptions] mockRequiredOnly:', serveOptions.mockRequiredOnly)
        logger.debug('[serveOptions] mockOptionalsAlways:', serveOptions.mockOptionalsAlways)
        logger.debug('[serveOptions] mockOptionalsProbability:', serveOptions.mockOptionalsProbability)
        logger.debug('[serveOptions] mockDataFileFirst:', serveOptions.mockDataFileFirst)
        logger.debug('[serveOptions] mockDataFileRootPath:', serveOptions.mockDataFileRootPath)

        // eslint-disable-next-line no-param-reassign
        projectDir = path.resolve(globalCmdOptions.cwd.value, projectDir || '')
        const globalContextConfig = calcGlobalContextConfig(projectDir)
        const contextConfig: ServeContextConfig = calcContextConfig(projectDir, {
          ...globalCmdOptions,
          ...serveOptions
        })

        // debug logger
        logger.debug('projectDir:', projectDir)
        logger.debug('[serveContextConfig] host:', contextConfig.host)
        logger.debug('[serveContextConfig] port:', contextConfig.port)
        logger.debug('[serveContextConfig] prefixUrl:', contextConfig.prefixUrl)
        logger.debug('[serveContextConfig] encoding:', contextConfig.encoding)
        logger.debug('[serveContextConfig] apiConfigPath:', contextConfig.apiConfigPath)
        logger.debug('[serveContextConfig] schemaRootPath:', contextConfig.schemaRootPath)
        logger.debug('[serveContextConfig] mockRequiredOnly:', contextConfig.mockRequiredOnly)
        logger.debug('[serveContextConfig] mockOptionalsAlways:', contextConfig.mockOptionalsAlways)
        logger.debug('[serveContextConfig] mockOptionalsProbability:', contextConfig.mockOptionalsProbability)
        logger.debug('[serveContextConfig] mockDataFileFirst:', contextConfig.mockDataFileFirst)
        logger.debug('[serveContextConfig] mockDataFileRootPath:', contextConfig.mockDataFileRootPath)

        const context = new RestfulApiToolServerContext(projectDir, globalContextConfig, contextConfig)
        const server = new RestfulApiToolServer(context)

        // execute hook before executing command
        for (const hook of self.beforeStartHooks) {
          await hook(server, context)
        }

        await server.start()

        // execute hook after command completed
        for (const hook of self.afterCompletedHooks) {
          await hook(server, context)
        }
      })
  }

  public onHook(type: SubCommandHook.AFTER_TRIGGERED, hook: ServeCommandAfterTriggeredHook): this
  public onHook(type: SubCommandHook.BEFORE_START, hook: ServeCommandBeforeStartHook): this
  public onHook(type: SubCommandHook.AFTER_COMPLETED, hook: ServeCommandAfterCompletedHook): this
  public onHook(type: SubCommandHook, hook: ServeCommandHook): this {
    switch (type) {
      case SubCommandHook.AFTER_TRIGGERED:
        this.afterTriggeredHooks.push(hook as ServeCommandAfterTriggeredHook)
        break
      case SubCommandHook.BEFORE_START:
        this.beforeStartHooks.push(hook as ServeCommandBeforeStartHook)
        break
      case SubCommandHook.AFTER_COMPLETED:
        this.afterCompletedHooks.push(hook as ServeCommandAfterCompletedHook)
        break
    }
    return this
  }

  public get defaultOptions(): ServeCmdOptions {
    const defaultOptions: ServeCmdOptions = {
      host: {
        value: '127.0.0.1',
        userSpecified: false,
      },
      port: {
        value: 8080,
        userSpecified: false,
        coverFunc: coverInteger,
      },
      mockRequiredOnly: {
        value: false,
        userSpecified: false,
      },
      mockOptionalsAlways: {
        value: false,
        userSpecified: false,
      },
      mockOptionalsProbability: {
        value: 0.8,
        userSpecified: false,
      },
      mockDataFileFirst: {
        value: false,
        userSpecified: false,
      },
      mockDataFileRootPath: {
        value: undefined,
        userSpecified: false,
        coverFunc: coverString as CoverOperationFunc<string | undefined>,
      },
    }
    return defaultOptions
  }
}
