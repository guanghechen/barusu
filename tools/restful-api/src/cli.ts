import { mountSubCommandGenerate } from './command/generate'
import { mountSubCommandInit } from './command/init'
import { mountSubCommandServe } from './command/serve'
import { createProgram } from './index'


const program = createProgram()


// load sub-command: init
mountSubCommandInit(program)

// load sub-command: generate
mountSubCommandGenerate(program)

// load sub-command: serve
mountSubCommandServe(program)


program
  .parse(process.argv)
