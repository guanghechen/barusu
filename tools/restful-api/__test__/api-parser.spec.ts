import fs from 'fs-extra'
import path from 'path'
import type { ResolvedApiItemGroup } from '../src'
import { ApiItemParser } from '../src'

// override
function formatApiItemGroup(data: ResolvedApiItemGroup[]): any {
  const projectDir = path.resolve()
  const stringifyFilter = (key: string, value: any): unknown => {
    // RegExp to string
    if (value instanceof RegExp) {
      return value.source
    }

    // replace absolute path with symbol
    if (typeof value === 'string') {
      if (value.startsWith(projectDir)) {
        return '<PROJECT_DIR>' + value.substr(projectDir.length)
      }
    }

    return value
  }
  return JSON.parse(JSON.stringify(data, stringifyFilter))
}

describe('api-parser', function () {
  const caseRootDirectory = path.resolve(__dirname, 'cases', 'api-parser')
  const kases = fs.readdirSync(caseRootDirectory)

  const parser = new ApiItemParser()
  for (const apiConfigFilePath of kases) {
    const title = path.parse(apiConfigFilePath).name
    // eslint-disable-next-line jest/valid-title
    test(title, async function () {
      const apiConfigFilepath = path.resolve(
        caseRootDirectory,
        apiConfigFilePath,
      )
      parser.scan(apiConfigFilepath, 'utf-8')
      const apiItemGroups = parser.collect()
      const data = formatApiItemGroup(apiItemGroups)
      expect(data).toMatchSnapshot('api items')
    })
  }
})
