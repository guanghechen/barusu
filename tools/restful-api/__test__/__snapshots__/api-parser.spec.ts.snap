// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-parser request-full-model: api items 1`] = `
Array [
  Object {
    "active": true,
    "desc": "",
    "fullName": "user",
    "items": Array [
      Object {
        "active": true,
        "desc": "",
        "method": "GET",
        "name": "me",
        "path": "/api/user/me",
        "request": Object {
          "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user/current-user-info-request-model.json",
          "voName": "CurrentUserInfoRequestModel",
        },
        "response": Object {
          "headers": Object {
            "Authorize": "<token>",
            "Content-Type": "application/json; UTF-8",
          },
          "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user/current-user-info-response-model.json",
          "voName": "CurrentUserInfoResponseModel",
        },
        "title": "获取当前用户信息",
      },
      Object {
        "active": true,
        "desc": "",
        "method": "GET",
        "name": "UserInfo",
        "path": "/api/user/:id",
        "request": Object {},
        "response": Object {
          "headers": Object {
            "Authorize": "<token>",
            "Content-Type": "application/json; UTF-8",
          },
          "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user/user-user-info-response-vo.json",
          "voName": "UserUserInfoResponseVo",
        },
        "title": "获取指定用户信息",
      },
    ],
    "name": "user",
    "path": "/api/user",
    "request": Object {
      "voNamePrefix": "",
      "voNameSuffix": "RequestVo",
    },
    "response": Object {
      "headers": Object {
        "Authorize": "<token>",
        "Content-Type": "application/json; UTF-8",
      },
      "voNamePrefix": "",
      "voNameSuffix": "ResponseVo",
    },
    "subGroups": Array [],
    "title": "user",
  },
]
`;

exports[`api-parser simple: api items 1`] = `
Array [
  Object {
    "active": true,
    "desc": "",
    "fullName": "user",
    "items": Array [
      Object {
        "active": true,
        "desc": "",
        "method": "POST",
        "name": "login",
        "path": "/api/user/login",
        "request": Object {},
        "response": Object {
          "headers": Object {
            "Content-Type": "application/json; UTF-8",
          },
          "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user/user-login-response-vo.json",
          "voName": "UserLoginResponseVo",
        },
        "title": "登录",
      },
      Object {
        "active": true,
        "desc": "",
        "method": "POST",
        "name": "logout",
        "path": "/api/user/logout",
        "request": Object {},
        "response": Object {
          "headers": Object {
            "Content-Type": "application/json; UTF-8",
          },
          "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user/user-logout-response-vo.json",
          "voName": "UserLogoutResponseVo",
        },
        "title": "退出登录",
      },
    ],
    "name": "user",
    "path": "/api/user",
    "request": Object {
      "voNamePrefix": "",
      "voNameSuffix": "RequestVo",
    },
    "response": Object {
      "headers": Object {
        "Content-Type": "application/json; UTF-8",
      },
      "voNamePrefix": "",
      "voNameSuffix": "ResponseVo",
    },
    "subGroups": Array [
      Object {
        "active": true,
        "desc": "",
        "fullName": "user-info",
        "items": Array [
          Object {
            "active": true,
            "desc": "",
            "method": "GET",
            "name": "me",
            "path": "/api/user/me",
            "request": Object {
              "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user-info/current-user-info-request-model.json",
              "voName": "CurrentUserInfoRequestModel",
            },
            "response": Object {
              "headers": Object {
                "Authorize": "<token>",
                "Content-Type": "application/json; UTF-8",
              },
              "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user-info/current-user-info-response-model.json",
              "voName": "CurrentUserInfoResponseModel",
            },
            "title": "获取当前用户信息",
          },
          Object {
            "active": true,
            "desc": "",
            "method": "GET",
            "name": "UserInfo",
            "path": "/api/user/:id",
            "request": Object {},
            "response": Object {
              "headers": Object {
                "Authorize": "<token>",
                "Content-Type": "application/json; UTF-8",
              },
              "schemaPath": "<PROJECT_DIR>/test/cases/schemas/user-info/xx-user-info-user-info-response-vo.json",
              "voName": "XxUserInfoUserInfoResponseVo",
            },
            "title": "获取指定用户信息",
          },
        ],
        "name": "info",
        "path": "/api/user",
        "request": Object {
          "voNamePrefix": "",
          "voNameSuffix": "RequestVo",
        },
        "response": Object {
          "headers": Object {
            "Authorize": "<token>",
            "Content-Type": "application/json; UTF-8",
          },
          "voNamePrefix": "Xx",
          "voNameSuffix": "ResponseVo",
        },
        "subGroups": Array [],
        "title": "info",
      },
    ],
    "title": "user",
  },
]
`;

exports[`api-parser user-blog: api items 1`] = `
Array [
  Object {
    "active": true,
    "desc": "basic api",
    "fullName": "user",
    "items": Array [
      Object {
        "active": true,
        "desc": "",
        "method": "GET",
        "name": "query",
        "path": "/api/user/me",
        "request": Object {},
        "response": Object {
          "schemaPath": "<PROJECT_DIR>/user/user-query-response-vo.json",
          "voName": "UserQueryResponseVo",
        },
        "title": "get current user information",
      },
      Object {
        "active": true,
        "desc": "",
        "method": "POST",
        "name": "update",
        "path": "/api/user/me",
        "request": Object {
          "schemaPath": "<PROJECT_DIR>/user/update-current-user-request-vo.json",
          "voName": "UpdateCurrentUserRequestVo",
        },
        "response": Object {
          "schemaPath": "<PROJECT_DIR>/user/update-current-user-response-vo.json",
          "voName": "UpdateCurrentUserResponseVo",
        },
        "title": "update current user information",
      },
      Object {
        "active": true,
        "desc": "",
        "method": "POST",
        "name": "login",
        "path": "/api/user/login",
        "request": Object {},
        "response": Object {
          "schemaPath": "<PROJECT_DIR>/user/user-login-response-vo.json",
          "voName": "UserLoginResponseVo",
        },
        "title": "login",
      },
    ],
    "name": "user",
    "path": "/api/user",
    "request": Object {
      "voNamePrefix": "",
      "voNameSuffix": "RequestVo",
    },
    "response": Object {
      "voNamePrefix": "",
      "voNameSuffix": "ResponseVo",
    },
    "subGroups": Array [],
    "title": "user",
  },
  Object {
    "active": true,
    "desc": "the user's blog",
    "fullName": "blog",
    "items": Array [
      Object {
        "active": true,
        "desc": "",
        "method": "GET",
        "name": "query-articles",
        "path": "/api/blog/articles",
        "request": Object {},
        "response": Object {
          "schemaPath": "<PROJECT_DIR>/blog/blog-query-articles-pagination-response-vo.json",
          "voName": "BlogQueryArticlesPaginationResponseVo",
        },
        "title": "query articles under the blog",
      },
      Object {
        "active": true,
        "desc": "",
        "method": "POST",
        "name": "create-article",
        "path": "/api/blog/articles",
        "request": Object {},
        "response": Object {
          "schemaPath": "<PROJECT_DIR>/blog/blog-create-article-pagination-response-vo.json",
          "voName": "BlogCreateArticlePaginationResponseVo",
        },
        "title": "create an article",
      },
    ],
    "name": "blog",
    "path": "/api/blog",
    "request": Object {
      "voNamePrefix": "",
      "voNameSuffix": "RequestVo",
    },
    "response": Object {
      "voNamePrefix": "",
      "voNameSuffix": "PaginationResponseVo",
    },
    "subGroups": Array [
      Object {
        "active": true,
        "desc": "blog articles",
        "fullName": "blog-article",
        "items": Array [
          Object {
            "active": true,
            "desc": "",
            "method": "GET",
            "name": "query",
            "path": "/api/blog/article/:articleId",
            "request": Object {},
            "response": Object {
              "schemaPath": "<PROJECT_DIR>/blog-article/blog-article-query-response-vo.json",
              "voName": "BlogArticleQueryResponseVo",
            },
            "title": "get article with specified \`articleId\`",
          },
          Object {
            "active": true,
            "desc": "",
            "method": "POST",
            "name": "update",
            "path": "/api/blog/article/:articleId",
            "request": Object {},
            "response": Object {
              "schemaPath": "<PROJECT_DIR>/blog-article/blog-article-update-response-vo.json",
              "voName": "BlogArticleUpdateResponseVo",
            },
            "title": "update article with specified \`articleId\`",
          },
          Object {
            "active": true,
            "desc": "",
            "method": "DELETE",
            "name": "delete",
            "path": "/api/blog/article/:articleId",
            "request": Object {},
            "response": Object {
              "schemaPath": "<PROJECT_DIR>/blog-article/blog-article-delete-response-vo.json",
              "voName": "BlogArticleDeleteResponseVo",
            },
            "title": "delete an article with specified \`articleId\`",
          },
        ],
        "name": "article",
        "path": "/api/blog/article/:articleId",
        "request": Object {
          "voNamePrefix": "",
          "voNameSuffix": "RequestVo",
        },
        "response": Object {
          "voNamePrefix": "",
          "voNameSuffix": "ResponseVo",
        },
        "subGroups": Array [],
        "title": "article",
      },
    ],
    "title": "blog",
  },
]
`;
